httpx/http_service.go

```diff
@@ -0,0 +1,165 @@
+package httpx
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/remind101/pkg/retry"
+
+	"golang.org/x/net/context"
+)
+
+type RoundTripper interface {
+	RoundTrip(*http.Request, context.Context) (*http.Response, error)
+}
+
+type Client struct {
+	innerClient		*http.Client
+	retryTransport 	*RetryTransport
+	baseURL 		*url.URL
+}
+
+type RetryTransport struct {
+	Retrier     *retry.Retrier
+	Transport	*http.Transport
+	Client		*Client
+}
+
+func NewHTTPClient(serviceName string, baseURL *url.URL) *Client {
+	retryTransport := NewRetryTransport(serviceName)
+
+	client := &Client{
+		innerClient: &http.Client{},
+		baseURL: baseURL,
+		retryTransport: retryTransport,
+	}
+	retryTransport.Client = client
+	client.innerClient.Transport = retryTransport.Transport
+
+	return client
+}
+func NewRetryTransport(serviceName string) *RetryTransport {
+	retrier := retry.NewErrorTypeRetrier(serviceName,
+		retry.DefaultBackOffOpts,
+		(*net.OpError)(nil),
+		(*RetryableHTTPError)(nil))
+
+	// DefaultTransport but with different numerical settings
+	transport := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		Dial: (&net.Dialer{
+			Timeout:   15 * time.Second,
+			KeepAlive: 90 * time.Second,
+		}).Dial,
+		TLSHandshakeTimeout: 3 * time.Second,
+	}
+
+	return &RetryTransport{
+		Retrier: retrier,
+		Transport: transport,
+	}
+}
+
+func (rt *RetryTransport) RoundTrip(req *http.Request, ctx context.Context) (*http.Response, error) {
+	ret, err := rt.Retrier.Retry(func() (interface{}, error) {
+		return rt.Client.call(req, ctx)
+	})
+	if err != nil {
+		return nil, err
+	}
+	if resp, ok := ret.(*http.Response); ok {
+		return resp, nil
+	} else {
+		panic("http.Response not of correct type")
+	}
+}
+
+func (client *Client) Do(req *http.Request, ctx context.Context) (*http.Response, error) {
+	return client.retryTransport.RoundTrip(req, ctx)
+}
+
+func (client *Client) call(req *http.Request, ctx context.Context) (*http.Response, error) {
+	req.Header.Add("X-Request-Id", RequestID(ctx))
```

Comment:
Doing stuff like this is better to do in a RoundTripper, so you can compose middleware easier:

```go
// RequestIDTransport is an http.RoundTripper implementation that adds the
// embedded request id to outgoing http requests.
type RequestIDTransport struct {
        Transport RoundTripper
}

func (t *RequestIDTransport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
        req.Header.Add("X-Request-Id", RequestID(ctx))
        return t.Transport.RoundTrip(ctx, req)
}
```